File: 2020/day/10/main.py
import fileinput
from collections import Counter

adapters = [int(x) for x in fileinput.input('input.txt')]
output = max(adapters) + 3


jolts = sorted([0, output] + adapters)
#print(ys)

d1, d3 = 0, 0

for i in range(len(jolts) - 1):
    d = jolts[i + 1] - jolts[i] 
    
    if d == 1:
        d1 += 1
    elif d == 3:
        d3 += 1

print(d1 * d3)

#
def part2(jolts):
    DP = {}

    def dp(i):
        if i == len(jolts) - 1:
            return 1

        if i in DP:
            return DP[i]
        
        n = 0
        j = i + 1

        while j < len(jolts) and jolts[j] - jolts[i] <= 3:
            n += dp(j)
            j += 1

        DP[i] = n
        return n

    return dp(0)
    
print(part2(jolts))

# low
# 12089663946752
# 13816758796288

File: 2020/day/10/Main.hs
import Data.List (sort)
import Data.Maybe
import qualified Data.IntMap as IntMap
import Data.IntMap (IntMap)

part1 jolts = 
  let deltas = zipWith (-) (tail jolts) jolts
  in length (filter (== 1) deltas) * length (filter (== 3) deltas)

part2 jolts = IntMap.lookup (last jolts) a
  where
    a = IntMap.fromListWith (+) [(i, f i) | i <- jolts]

    f 0 = 1
    f i = IntMap.findWithDefault 0 (i - 3) a
        + IntMap.findWithDefault 0 (i - 2) a
        + IntMap.findWithDefault 0 (i - 1) a

main = do
    input <- readFile "input.txt"
    let adapters = map read (lines input)
        jolts = 0 : sort adapters ++ [maximum adapters + 3]

    -- Part 1
    print $ part1 jolts

    -- Part 2
    print $ fromJust $ part2 jolts